frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -1305,6 +1305,114 @@ const IR::Node* TypeInference::postorder(IR::Concat* expression) {
     return expression;
 }
 
+const IR::Node* TypeInference::postorder(IR::Key* key) {
+    LOG1("TypeInference: visiting " << dbp(key));
+    // compute the type and store it in typeMap
+    auto vec = new IR::Vector<IR::Type>();
+    for (auto ke : *key->keyElements) {
+        auto kt = typeMap->getType(ke->expression);
+        if (kt == nullptr) {
+            LOG2("Bailing out for " << dbp(ke));
+            return key;
+        }
+        vec->push_back(kt);
+    }
+    auto keyTuple = new IR::Type_Tuple(Util::SourceInfo(), vec);
+    LOG2("Setting key type to " << dbp(keyTuple));
+    setType(key, keyTuple);
+    // installing also for the original because we cannot tell which one will survive in the ir
+    // \TODO: figure out the rules of survivorship
+    LOG2("Setting key type to " << dbp(getOriginal()));
+    setType(getOriginal(), keyTuple);
+    return key;
+}
+
+/**
+    typecheck a table initializer entry list
+
+    The invariants are:
+    - table keys must have been type checked before entries
+*/
+const IR::Node* TypeInference::postorder(IR::EntriesList* el) {
```

Comment:
I think that adding these checks here is too late. In postorder this method is visited after the IR::Entry method, which will have failed by now. So these checks have to be moved there. 

Alternatively, we can check that keys and actions come first in a separate pass, perhaps the early validation pass.